### DISCUSSION TITLE
Custom Script

### DISCUSSION THREAD

**Discussion by RenatoPereira91** [netbox-community/netbox#13976] [General]
Hello everyone,

Today I have a script with Python that I can add Cisco device in the NetBox, I'm trying to convert to Custom Script  to make it easier to interact (today I run the python and setting the information to start).

But now I have one doubt, I'm trying to add one interface to device type but never work I getting the err in the moment "new_intf.save(). I'm post my code below ":

-------- CODE --------

                      new_device_type = DeviceType(
                          manufacturer=data['manufacturer'],
                          model=structured_inventory["slot"][device_inventory_ID]["rp"][device_inventory]['pid'],
                          slug=structured_inventory["slot"][device_inventory_ID]["rp"][device_inventory]['pid'],
                          part_number=structured_inventory["slot"][device_inventory_ID]["rp"][device_inventory]['descr'],
                          is_full_depth=0,
                          comments="Added via API"
                      )
                      new_device_type.save()
                      self.log_info(f"Created the new device: {new_device_type}")
                 
                      for interface in structured_interfaces.keys():
                          filterhardinterface = re.compile("^[gGeEfFtT]\S+[0-9]/?[0-9]*")
                          filterPrimeraInterface = re.compile("^[gGeEfFtT]\S+"+device_inventory_ID+"/0/?[0-9]*")
                          if filterhardinterface.search(interface):
                              if device_inventory_ID=="1":
                                  if "." not in interface:
                                      new_intf = Interface(
                                          name=interface,
                                          type=Interface_Type[structured_interfaces[interface]['media_type']],
                                          device_id=new_device_type
                                      )
                                      
                              else:
                                  if filterPrimeraInterface.search(interface) and "." not in interface:                 
                                      interface1 = interface.split("/")
                                      interface = interface1[0].rstrip(interface1[0][-1])+"1/"+interface1[1]+"/"+interface1[2]
                                      new_intf = Interface(
                                          name=interface,
                                          devicetype=new_device_type,
                                          type=Interface_Type[structured_interfaces[interface]['media_type']],
                                      )
                              new_intf.save()
                              new_device_type.assigned_object = new_intf

**Comment by mathewbishop** [comment]
Hey @RenatoPereira91!

I notice when you define `new_intf` in your code, you are only defining it if certain conditions are met. 

When you call `new_intf.save()`, you're calling it outside of any of the conditional statements where `new_intf` could possibly be defined. This opens up the possibility that the call to save the interface might fail because the conditions are not being met in which `new_intf` gets defined, and thus `new_intf.save()` is getting called on an undefined variable (`new_itnf`) in some cases. 

One alternative might be to include the call to save `new_intf` inside the respective conditions where it is defined. This will help to eliminate that particular possibility that `new_intf` is not defined when `.save()` is called. 

Something like the following:


```py
               for interface in structured_interfaces.keys():
                      filterhardinterface = re.compile("^[gGeEfFtT]\S+[0-9]/?[0-9]*")
                      filterPrimeraInterface = re.compile("^[gGeEfFtT]\S+"+device_inventory_ID+"/0/?[0-9]*")
                      if filterhardinterface.search(interface):
                          if device_inventory_ID=="1":
                              if "." not in interface:
                                  new_intf = Interface(
                                      name=interface,
                                      type=Interface_Type[structured_interfaces[interface]['media_type']],
                                      device_id=new_device_type
                                  )
                                  new_intf.full_clean() # best practice to call full_clean() first
                                  new_intf.save()
                                  new_device_type.assigned_object = new_intf
                          else:
                              if filterPrimeraInterface.search(interface) and "." not in interface:                 
                                  interface1 = interface.split("/")
                                  interface = interface1[0].rstrip(interface1[0][-1])+"1/"+interface1[1]+"/"+interface1[2]
                                  new_intf = Interface(
                                      name=interface,
                                      devicetype=new_device_type,
                                      type=Interface_Type[structured_interfaces[interface]['media_type']],
                                  )
                                  new_intf.full_clean() # best practice to call full_clean() first
                                  new_intf.save()
                                  new_device_type.assigned_object = new_intf
```

<br/>

Also, noticed I have added `full_clean()` calls before each save call. From what I've read, this is best practice, and I believe it has to do with some legacy, backward-compatibility concerns with the implementation of the Django ORM. 



<br/>
<br/>

Hope this helps! If you you're still having trouble, it would be helpful to post the specific error message you're getting.

**Reply by RenatoPereira91** [reply]
Hi @mathewbishop, I tried your config, but I get this wrong:

An exception occurred: ValueError: Cannot assign "<DeviceType: WS-C2960-8TC-L>": "Interface.device" must be a "Device" instance.

Traceback (most recent call last):
  File "/opt/netbox/netbox/extras/scripts.py", line 504, in _run_script
    script.output = script.run(data=data, commit=commit)
  File "/opt/netbox/netbox/scripts/Add Device.py", line 169, in run
    createdevicetype(structured_inventory,structured_interfaces)
  File "/opt/netbox/netbox/scripts/Add Device.py", line 145, in createdevicetype
    new_intf = Interface(
  File "/opt/netbox/netbox/dcim/models/device_components.py", line 86, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/base.py", line 543, in __init__
    _setattr(self, field.name, rel_obj)
  File "/opt/netbox/netbox/utilities/tracking.py", line 78, in __setattr__
    super().__setattr__(name, value)
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/fields/related_descriptors.py", line 266, in __set__
    raise ValueError(
ValueError: Cannot assign "<DeviceType: WS-C2960-8TC-L>": "Interface.device" must be a "Device" instance.

I tried the comment the device_id, but this field is mandatory:

An exception occurred: ValidationError: {'device': ['This field cannot be null.']}

Traceback (most recent call last):
  File "/opt/netbox/netbox/extras/scripts.py", line 504, in _run_script
    script.output = script.run(data=data, commit=commit)
  File "/opt/netbox/netbox/scripts/Add Device.py", line 168, in run
    createdevicetype(structured_inventory,structured_interfaces)
  File "/opt/netbox/netbox/scripts/Add Device.py", line 149, in createdevicetype
    new_intf.full_clean() # best practice to call full_clean() first
  File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/base.py", line 1502, in full_clean
    raise ValidationError(errors)
django.core.exceptions.ValidationError: {'device': ['This field cannot be null.']}

**Comment by mathewbishop** [comment]
Ah okay, are you trying to create a new device or a new device type? 

The error is indicating that you cannot pass a `DeviceType` object to the `device` parameter when creating an `Interface` object. So in this case I'm guessing you may be wanting to create a Device object instead, and then passing that to the `device` parameter when creating an interface. 


This is creating a Device Type, not a Device:

```py
new_device_type = DeviceType(
                      manufacturer=data['manufacturer'],
                      model=structured_inventory["slot"][device_inventory_ID]["rp"][device_inventory]['pid'],
                      slug=structured_inventory["slot"][device_inventory_ID]["rp"][device_inventory]['pid'],
                      part_number=structured_inventory["slot"][device_inventory_ID]["rp"][device_inventory]['descr'],
                      is_full_depth=0,
                      comments="Added via API"
                  )
                  new_device_type.save()
                  self.log_info(f"Created the new device: {new_device_type}")
```

So when you pass this object to the `device` parameter on the `Interface` object, it doesn't accept it because it's the wrong type of object for that parameter. 

<br/>
<hr/>
<br/>


To create a Device, you'd do something like:

```py
new_device = Device(
            name=name,
            device_type=type,
            device_role=role,
            tenant=tenant,
            platform=platform,
            site=site,
            custom_field_data=custom_fields,
            config_template_id=config_template_id,
            rack=rack,
            position=position,
            face=face,
        )
        new_device.full_clean()
        new_device.save()
        self.log_info(f"Created the new device: {new_device}")
```

<br/>

Then, to create an interface on that device:


```py
        iface = Interface(
            device=new_device,
            name=name,
            label=label,
            type=type,
            mgmt_only=is_mgmt_only,
            description=description,
        )
        iface.full_clean()
        iface.save()
        self.log_success(f"Created interface")
```

Here we are passing the device object `new_device` as the `device` parameter on the new `Interface` object.


<br/>
<hr/>
<br/>

If you were in fact trying to create a new DeviceType and then create interface templates on that DeviceType, you would need to find the specific models and parameters to use for that purpose. 

Netbox is architectured as a collection of data object types and relationships between them. You can explore these relationships by using the Netbox Shell. See @candlerb's comment [here](https://github.com/netbox-community/netbox/discussions/10534#discussioncomment-3782460) to learn more (also try searching for _nbshell_ in the discussions on GitHub, there are many other such comments and discussions that mention the Netbox Shell that you may find helpful). 

It's all about finding which objects you need, how they relate to other objects, and what kinds of parameters those objects are expecting in order to interact with them. 

The Netbox API documentation can also be helpful as a place to start. You'll still probably need the Netbox Shell to get more familiar with the data models that exist, but the API is exposing the same underlying data models, so sometimes it can give you a rough idea of what you're looking for as you dive into the nbshell.

**Reply by RenatoPereira91** [reply]
Hello,

Today I have the python code that run this:

1 - Collect all information about device
2 - Check if have device type, if not, create with physical interfaces
3 - Check IPAM with all IPs in device
4 - Create device
5 - Update all interfaces with description, plan, status, port channel, interface plan.

This script I run via cli, but my ideia is convert to use in custom script.

**Reply by candlerb** [reply]
From the Netbox online documentation [here](https://docs.netbox.dev/en/stable/), click on "Data model", "DCIM" and you'll find a brief mention of InterfaceTemplate [here](https://docs.netbox.dev/en/stable/models/dcim/interfacetemplate/).

In the source code, the DCIM models can be found [here](https://github.com/netbox-community/netbox/tree/v3.6.3/netbox/dcim/models) - see `devices.py`, `device_components.py` and `device_component_templates.py`

[InterfaceTemplate is to DeviceType](https://github.com/netbox-community/netbox/blob/v3.6.3/netbox/dcim/models/device_component_templates.py#L112-L118) as [Interface is to Device](https://github.com/netbox-community/netbox/blob/v3.6.3/netbox/dcim/models/device_components.py#L49-L53).

Creating DeviceTypes dynamically from scraped information may end up with inaccurate results (e.g. it will be hard to get modules/module bays correct, similarly for power supplies and console ports).  Another option would be to import the DeviceType from the [device type library](https://github.com/netbox-community/devicetype-library), if it already exists.
